### PR TITLE
Gutenberg Data Tutorial 5: Adding a Delete Button

### PR DESCRIPTION
**Pull request by adamziel**
### Description

This PR introduces the 5th part of [Create your First App with Gutenberg Data](https://github.com/WordPress/gutenberg/pull/38250) tutorial – the focus is on deleting entity records.

Follows up on the fourth part: https://github.com/WordPress/gutenberg/pull/39261 (sadly, the images aren't rendered properly there)

## See the [rich-text preview](https://github.com/WordPress/gutenberg/blob/dbc73deea57274d3f4f61d53c605ee210b6745d0/docs/how-to-guides/data-basics/5-adding-a-delete-button.md)

### CODE REVIEW

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 1]
```diff
@@ -0,0 +1,333 @@
...
+
+All the pieces are in place, great! Here’s all the changes we've made in this chapter:
+
+```js
+import { useDispatch } from '@wordpress/data';
+import { useEffect } from '@wordpress/element';
+import { Button, Modal, TextControl } from '@wordpress/components';
+
+function MyFirstApp() {
+	const [searchTerm, setSearchTerm] = useState( '' );
```
Comment (ryanwelcher) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md:316]: Small nitpick but we should add `useState` to the imports above.

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 2]
```diff
@@ -0,0 +1,333 @@
...
+	const { pages, hasResolved } = useSelect(
```
Comment (ryanwelcher) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md:317]: Same with `useSelect`

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md (hunk 1 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/5-adding-a-delete-button.md:316]: I initially didn't because it's imported in the previous part of the tutorial. Now that I think about it again, it would make more sense to make this behavior explicit 👍

Comment (adamziel) [reply docs/how-to-guides/data-basics/5-adding-a-delete-button.md:316]: Done @ryanwelcher

**Comment by adamziel** [comment]
cc @getdave @scruffian @draganescu

**Comment by scruffian** [comment]
This looks good. A couple of ideas:

1. Hiding the item before the delete action has succeeded feels like a non-ideal pattern - what if the delete fails? What about changing the visual appearance of the line, rather than hiding it?

2. It would be good to have a screencast of the whole flow at the end.

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md [hunk 3]
```diff
@@ -273,10 +273,11 @@ A few messages like  _Loading…_ or _No results_ would clear it up. Let’s imp
 ```js
 import { SearchControl, Spinner } from '@wordpress/components';
 function PagesList( { hasResolved, pages } ) {
-	if ( !hasResolved ) {
-		return <Spinner/>
-	}
 	if ( !pages?.length ) {
+		if ( !hasResolved ) {
+			return <Spinner/>
+		}
```
Comment (dmsnell) [inline docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]: I find this code change strange unless we're going to come back later in the tutorial and clean it up. We took what were separate concepts (`no results` and `no results yet`) and mashed them together by nesting the `if's.

Did we intend on doing this? What was the problem with the early-abort style it had already?

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 4]
```diff
@@ -0,0 +1,336 @@
...
+		if ( error ) {
+			alert( error );
+		}
+	}, [error] )
+
+	// ...
+}
+```
+
+Great! `DeletePageButton` is now fully aware of errors (even if `alert` is not the prettiest way to communicate them).
```
Comment (dmsnell) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: I love how we're demonstrating that we can get the error if the request to delete fails, but I almost feel like this is too much of a behavior we want to avoid to include it in the official docs. The attached screenshot tells the story enough on its own I think: there's no useful or actionable information present in that message and it's impact will be spreading fear when nobody can do anything about it.

If we want to show error handling I have these questions:
 - what kind of errors should we expect to get? not conceptually, but what is the return value of `getLastEntityDeleteError`? not great that we recreated C-style "get last error" calling patterns, but if we do that, can we at least expect some error codes or stable strings?
 - how would we ideally communicate the error to the person at the screen? this doc has great examples of using spinners and indicators to communicate where in the process we are, but for errors we are kinda just dropping the ball and giving up. could we use a Toast message of some kind to indicate the failure?
 - how should we respond after failure? retry? disable some control? even a helpful string message as used on invalid form fields could provide more actionable information.

if these docs didn't have the power to establish patterns and guide people's habits I wouldn't consider this as big of a deal, but I think we should write these with the idea that whatever we put out there is what we are telling the community we want to see. it would be worth removing this section or investing a little more in it to set a good example.

File: docs/how-to-guides/data-basics/1-data-basics-setup.md [hunk 5]
```diff
@@ -40,37 +40,56 @@ window.addEventListener(
 
 ```css
 .toplevel_page_my-first-gutenberg-app #wpcontent {
-	background: #fff;
+	background: #FFF;
```
Comment (getdave) [inline docs/how-to-guides/data-basics/1-data-basics-setup.md]:
```suggestion
	background: #fff;
```

Why the change?

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md (hunk 3 above)
Comment (getdave) [reply docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]: What he said 👆

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 6]
```diff
@@ -0,0 +1,437 @@
+# Adding a delete button
+
+This part is about adding a *Delete* feature to our app. Here's a glimpse of what we're going to build:
```
Comment (getdave) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]:
```suggestion
In this section we're going to be adding a *Delete* feature to our app. Here's a glimpse of what we're going to build:
```

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 7]
```diff
@@ -0,0 +1,437 @@
...
+						</td>
+					</tr>
+				) ) }
+			</tbody>
+		</table>
+	);
+}
+```
+
+This is how the PagesList should look like now:
```
Comment (getdave) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]:
```suggestion
This is what the PagesList should look like now:
```

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 8]
```diff
@@ -0,0 +1,437 @@
...
+
+![](https://raw.githubusercontent.com/WordPress/gutenberg/HEAD/docs/how-to-guides/data-basics/media/delete-button/deleting-in-progress.png)
+
+### Step 4: Handle errors
+
+We optimistically assumed that a *delete* operation would always succeed. Unfortunately, under the hood, it is a REST API request that can fail in many ways:
+
+* The website can be down
+* The update may be invalid
+* The page could have been deleted by someone else in the meantime
```
Comment (getdave) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]:
```suggestion
* The website can be down.
* The update may be invalid.
* The page could have been deleted by someone else in the meantime.
```

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 9]
```diff
@@ -0,0 +1,437 @@
...
+
+The `error` object comes from the `@wordpress/api-fetch` and contains information about the failure reason. It has the following properties:
+
+* `message` – a human-readable error message such as `Invalid post ID`.
+* `code` – a string-based error code such as `rest_post_invalid_id`. To learn about all possible error codes you'd need to refer to the [`/v2/pages` endpoint's source code](https://github.com/WordPress/wordpress-develop/blob/2648a5f984b8abf06872151898e3a61d3458a628/src/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php#L226-L230).
+* `data` (optional) – error details, contains the `code` property containing the HTTP response code for the failed request.
+
+In this tutorial, we will display the `error.message` to the user.
+
+In WordPress, there is an established patterns of displaying status information using the `Snackbar` component:
```
Comment (getdave) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]:
```suggestion
In WordPress, there is an established pattern of displaying status information using the `Snackbar` component:
```

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 10]
```diff
@@ -0,0 +1,437 @@
...
+			createErrorNotice( , {
+				type: 'snackbar',
+			} );
+		}
+	}, [ error ] )
+    // ...
+}
+```
+
+Great! `DeletePageButton` is now fully aware of errors. Let's see that error message in action. We'll trigger an invalid delete and let it fail. One way to do it is to set the `id` to `-1`:
```
Comment (getdave) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]:
```suggestion
Great! `DeletePageButton` is now fully aware of errors. Let's see that error message in action. We'll trigger an invalid delete and let it fail. One way to do this is to set the `id` to `-1`:
```

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 11]
```diff
@@ -0,0 +1,437 @@
...
+
+```js
+export function DeletePageButton( { pageId, onCancel, onSaveFinished } ) {
+	// ...
+	pageId = pageId * -1;
```
Comment (getdave) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: Nit: this is actually negating the ID (i.e. `8` would become `-8` rather than setting to `-1` as you say above 🤔

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 12]
```diff
@@ -0,0 +1,437 @@
...
+function DeletePageButton( { pageId } ) {
+	const { deleteEntityRecord } = useDispatch( coreDataStore );
+	const { createSuccessNotice, createErrorNotice } = useDispatch( noticesStore );
+	const handleDelete = async () => {
+		const success = await deleteEntityRecord( 'postType', 'page', pageId );
+		if ( success ) {
+			createSuccessNotice( "The page was deleted!", {
+				type: 'snackbar',
+			} );
+		}
```
Comment (getdave) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: This does feel like an unusual pattern. 

The "success" status is checked by awaiting the result of the Promise but the error case is handled somewhere completely different. 

If I was looking at this in isolation I'd be thinking that the developer isn't handling the error state.

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 13]
```diff
@@ -0,0 +1,437 @@
...
+		( select ) => {
+			const query = {};
+			if ( searchTerm ) {
+				query.search = searchTerm;
+			}
+			const selectorArgs = ['postType', 'page', query];
+			const pages = (
+				select( coreDataStore )
+					.getEntityRecords( ...selectorArgs )
+					?.filter( page => ! select( coreDataStore ).isDeletingEntityRecord('postType', 'page', page.id ) )
```
Comment (getdave) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: This is making sure we're not displayed any pages currently being deleted right?

**Review by getdave — APPROVED** [review]
Great work as always. Easy to read and to follow. Left some nits.

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md (hunk 13 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: Good spot! I just removed it.

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md (hunk 12 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: I just fixed that, hopefully it rings better now!

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md (hunk 11 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: I just updated the description.

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md (hunk 4 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: Good point! I followed the same pattern as the rest of WordPress and added Snackbar displaying the `error.message`.

File: docs/how-to-guides/data-basics/2-building-a-list-of-pages.md (hunk 3 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/2-building-a-list-of-pages.md]: I'm not sure what was my intention originally. I just reverted that change.

**Comment by adamziel** [comment]
I addressed all the feedback, how do you like it after this one last round of changes @getdave? Should we ship it?

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 14]
```diff
@@ -0,0 +1,454 @@
...
+				</tr>
+			</thead>
+			<tbody>
+				{ pages?.map( ( page ) => (
+					<tr key={page.id}>
+						<td>{ decodeEntities( page.title.rendered ) }</td>
+						<td>
+							<div className="form-buttons">
+								<EditPageButton pageId={ page.id } />
+								<DeletePageButton pageId={ page.id } />
```
Comment (draganescu) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: It would be cool if we had a way to highlight the addition somehow.

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 15]
```diff
@@ -0,0 +1,454 @@
...
+
+This is what the PagesList should look like now:
+
+![](https://raw.githubusercontent.com/WordPress/gutenberg/HEAD/docs/how-to-guides/data-basics/media/delete-button/delete-button.png)
+
+### Step 2: Wire the button to a delete action
+
+In Gutenberg data, we delete entity records from the WordPress REST API using the `deleteEntityRecord` action. It sends the request, processes the result, and updates the cached data in the Redux state.
+
+Here's how you can try it in your browser's dev tools:
```
Comment (draganescu) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]:
```suggestion
Here's how you can try deleting entity records in your browser's dev tools:
```

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 16]
```diff
@@ -0,0 +1,454 @@
...
+// We need a valid page ID to call deleteEntityRecord, so let's get the first available one using getEntityRecords.
+const pageId = wp.data.select( 'core' ).getEntityRecords( 'postType', 'page' )[0].id;
+
+// Now let's delete that page:
+const promise = wp.data.dispatch( 'core' ).deleteEntityRecord( 'postType', 'page', pageId );
+
+// promise gets resolved or rejected when the API request succeeds or fails.
+```
+
+Once the REST API request is finished, you will notice one of the pages has disappeared from the list. This is because that list is populated by using `useSelect()` and the `select( coreDataStore ).getEntityRecords( 'postType', 'page' )` selector. Anytime the underlying data changes, the list gets re-rendered with fresh data. That's pretty convenient!
```
Comment (draganescu) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]:
```suggestion
Once the REST API request is finished, you will notice one of the pages has disappeared from the list. This is because that list is populated by the `useSelect()` hook and the `select( coreDataStore ).getEntityRecords( 'postType', 'page' )` selector. Anytime the underlying data changes, the list gets re-rendered with fresh data. That's pretty convenient!
```

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md (hunk 14 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: I know, right? Maybe I could add a comment like 

``` This is the addition v```

Or! Use the `diff` format 

```diff
	<div className="form-buttons">
		<EditPageButton pageId={ page.id } />
+		<DeletePageButton pageId={ page.id } />
```

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 17]
```diff
@@ -0,0 +1,454 @@
...
+			isDeleting: select( coreDataStore ).isDeletingEntityRecord( 'postType', 'page', pageId ),
+		}),
+		[ pageId ]
+	)
+	return (
+		<Button variant="primary" onClick={ handleDelete } disabled={ isDeleting }>
+			{ isDeleting ? (
+				<>
+					<Spinner />
+					Deleting...
```
Comment (draganescu) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md:116]: Maybe for the sake of good practice we should use `__( string )` for strings? Maybe it's too much for this tutorial tho.

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 18]
```diff
@@ -0,0 +1,454 @@
...
+Here's what it looks like in action:
+
+![](https://raw.githubusercontent.com/WordPress/gutenberg/HEAD/docs/how-to-guides/data-basics/media/delete-button/deleting-in-progress.png)
+
+### Step 4: Handle errors
+
+We optimistically assumed that a *delete* operation would always succeed. Unfortunately, under the hood, it is a REST API request that can fail in many ways:
+
+* The website can be down.
+* The update may be invalid.
```
Comment (draganescu) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: What update?

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 19]
```diff
@@ -0,0 +1,454 @@
...
+* The page could have been deleted by someone else in the meantime.
+
+To tell the user when any of these happens, we need to extract the error information using the `getLastEntityDeleteError` selector:
```
Comment (draganescu) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]:
```suggestion
To tell the user when any of these errors happen, we need to extract the error information using the `getLastEntityDeleteError` selector:
```

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 20]
```diff
@@ -0,0 +1,454 @@
...
+
+```js
+// Replace 9 with an actual page ID
```
Comment (draganescu) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md:139]: This is a bit confusing. Why not use a placeholder instead of this comment about the magic number?

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 21]
```diff
@@ -0,0 +1,454 @@
...
+		if ( error ) {
+			// Display the error
+		}
+	}, [error] )
+
+	// ...
+}
+```
+
+The `error` object comes from the `@wordpress/api-fetch` and contains information about the failure reason. It has the following properties:
```
Comment (draganescu) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]:
```suggestion
The `error` object comes from the `@wordpress/api-fetch` and contains information about the error. It has the following properties:
```

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 22]
```diff
@@ -0,0 +1,454 @@
...
+
+* `message` – a human-readable error message such as `Invalid post ID`.
+* `code` – a string-based error code such as `rest_post_invalid_id`. To learn about all possible error codes you'd need to refer to the [`/v2/pages` endpoint's source code](https://github.com/WordPress/wordpress-develop/blob/2648a5f984b8abf06872151898e3a61d3458a628/src/wp-includes/rest-api/endpoints/class-wp-rest-revisions-controller.php#L226-L230).
+* `data` (optional) – error details, contains the `code` property containing the HTTP response code for the failed request.
+
+There are many ways to turn that data into an error message, but in this tutorial, we will display the `error.message`.
```
Comment (draganescu) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]:
```suggestion
There are many ways to turn that object into an error message, but in this tutorial, we will display the `error.message`.
```

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md (hunk 17 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/5-adding-a-delete-button.md:116]: @scruffian have already mentioned that – I'd like to spin another PR to update all the parts of this tutorial at once.

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md (hunk 18 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: Good catch! I copied that from the "update" part of the tutorial :D

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 23]
```diff
@@ -0,0 +1,454 @@
...
+
+WordPress has an established pattern of displaying status information using the `Snackbar` component. Here's what it looks like **in the Widgets editor**:
```
Comment (draganescu) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md:174]: Why the emphasis?

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md (hunk 20 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/5-adding-a-delete-button.md:139]: How would you introduce a placeholder that would clearly communicate that?

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md (hunk 23 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/5-adding-a-delete-button.md:174]: The screenshot isn't from the plugin we're building so I'm hoping to avoid confusion "wait, what is this screen?"

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md (hunk 14 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: Hm on a second thought, the handbook probably wouldn't correctly render a diff.

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 24]
```diff
@@ -0,0 +1,454 @@
...
+Let's use the same type of notifications in our plugin! There are two parts to this:
+
+1. Displaying notifications
+2. Dispatching notifications
+
+#### Displaying notifications
+
+Our application only knows how to display pages but does not know how to display notifications. Let's tell it!
+
+The [`Snackbar` WordPress component](https://wordpress.github.io/gutenberg/?path=/story/components-snackbar--default) represents a single notification:
```
Comment (draganescu) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: The jump from let's tell it! to here are two new components seems to sudden. Maybe a sentence about how we can use these components such and such would do, and then explain what eachone does.

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 25]
```diff
@@ -0,0 +1,454 @@
...
+	const { getLastEntityDeleteError } = useSelect( coreDataStore )
+	const handleDelete = async () => {
+		const success = await deleteEntityRecord( 'postType', 'page', pageId);
+		if ( success ) {
+			// Tell the user the operation succeeded:
+			createSuccessNotice( "The page was deleted!", {
+				type: 'snackbar',
+			} );
+		} else {
+			// We use the selector directly to get the error at this point in time.
```
Comment (draganescu) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: This still mysterious and I think the comment makes things worse :)

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md [hunk 26]
```diff
@@ -0,0 +1,454 @@
...
+			// Imagine we fetched the error like this:
+			//     const { lastError } = useSelect( function() { /* ... */ } );
+			// Then, lastError would be null inside of handleDelete.
+			// Why? Because we'd refer to the version of it that was computed
+			// before the handleDelete was even called.
+			const lastError = getLastEntityDeleteError( 'postType', 'page', pageId );
+			const message = ( lastError?.message || 'There was an error.' ) + ' Please refresh the page and try again.'
+			// Tell the user how exactly the operation have failed:
```
Comment (draganescu) [inline docs/how-to-guides/data-basics/5-adding-a-delete-button.md]:
```suggestion
			// Tell the user how exactly the operation has failed:
```

**Review by draganescu — APPROVED** [review]
Left a few nits but this is awesome otherwise!

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md (hunk 25 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: I agree, I replaced it with

```
			// We use the selector directly to get the fresh error *after* the deleteEntityRecord
			// have failed.
```

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md (hunk 24 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: Good point, added!

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md (hunk 14 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/5-adding-a-delete-button.md]: I went for
```
{/* ↓ This is the only change in the PagesList component */}
<DeletePageButton pageId={ page.id }/>
{/* ↑ This is the only change in the PagesList component */}
```

File: docs/how-to-guides/data-basics/5-adding-a-delete-button.md (hunk 20 above)
Comment (adamziel) [reply docs/how-to-guides/data-basics/5-adding-a-delete-button.md:139]: I left the placeholder for now, I'm happy to revisit later on, though!

**Comment by dmsnell** [comment]
Thanks for this great addition. One thing I noted is that the `error` handling code flashes an `alert` with `useEffect( …, [error])` and I think that means we should expect an empty alert window when the error disappears. Might be best to try and omit any examples with `alert()` in them.

Alternatively we could use that to render a modal dialog, issue a toast when truthy, render an error message, or simply omit the rendering code for the example purposes.

Other than that I think these tutorials are getting so much more valuable 👍

**Comment by adamziel** [comment]
@dmsnell Hm I was sure I removed the `alert()` calls from this PR in favor of the snackbar notifications. Unless you mean this part still talking about `useEffect`?

```js
const DeletePageButton = ({ pageId }) => {
	// ...
	const { error, /* ... */ } = useSelect(
		select => ( {
			error: select( coreDataStore ).getLastEntityDeleteError( 'postType', 'page', pageId ),
			// ...
		} ),
		[pageId]
	);
	useEffect( () => {
		if ( error ) {
			// Display the error
		}
	}, [error] )
	// ...
}
```

If so – great spot, I'll rework it to be consistent with the error handling code later in the same document.

**Comment by dmsnell** [comment]
if you removed it I might have missed it because I followed the link in the description, which referenced a specific commit, so maybe it referenced an earlier version of the PR.

the main thing was in that `useEffect` block when the error disappears the code runs again so as long as we don't make it seem like it only runs when there _is_ an error it's all good. you may have already done this.

**Comment by adamziel** [comment]
It's already taken care of @dmsnell:

```js
import { store as noticesStore } from '@wordpress/notices';
import { useEffect } from '@wordpress/element';
function DeletePageButton( { pageId } ) {
	const { createSuccessNotice, createErrorNotice } = useDispatch( noticesStore );
	// useSelect returns a list of selectors if you pass the store handle
	// instead of a callback:
	const { getLastEntityDeleteError } = useSelect( coreDataStore )
	const handleDelete = async () => {
		const success = await deleteEntityRecord( 'postType', 'page', pageId);
		if ( success ) {
			// Tell the user the operation succeeded:
			createSuccessNotice( "The page was deleted!", {
				type: 'snackbar',
			} );
		} else {
			// We use the selector directly to get the fresh error *after* the deleteEntityRecord
			// have failed.
			const lastError = getLastEntityDeleteError( 'postType', 'page', pageId );
			const message = ( lastError?.message || 'There was an error.' ) + ' Please refresh the page and try again.'
			// Tell the user how exactly the operation has failed:
			createErrorNotice( message, {
				type: 'snackbar',
			} );
		}
	}
	// ...
}
```
